cmd/promtool/unittest.go

```diff
@@ -0,0 +1,467 @@
+		return []error{err}
+	}
+
+	if unitTestInp.EvaluationInterval == 0 {
+		unitTestInp.EvaluationInterval = 1 * time.Minute
+	}
+
+	// Bounds for evaluating the rules.
+	mint := time.Unix(0, 0)
+	maxd := unitTestInp.maxEvalTime()
+	maxt := mint.Add(maxd)
+	// Rounding off to nearest Eval time (> maxt).
+	maxt = maxt.Add(unitTestInp.EvaluationInterval / 2).Round(unitTestInp.EvaluationInterval)
+
+	// Giving number for groups mentioned in the file for ordering.
+	// Lower number group should be evaluated before higher number group.
+	groupOrderMap := make(map[string]int)
+	for i, gn := range unitTestInp.GroupEvalOrder {
+		if _, ok := groupOrderMap[gn]; ok {
+			return []error{fmt.Errorf("Group name repeated in evaluation order: %s", gn)}
+		}
+		groupOrderMap[gn] = i
+	}
+
+	// Testing.
+	var errs []error
+	for _, t := range unitTestInp.Tests {
+		ers := t.test(mint, maxt, unitTestInp.EvaluationInterval, groupOrderMap,
+			unitTestInp.RuleFiles...)
+		if ers != nil {
+			errs = append(errs, ers...)
+		}
+	}
+
+	if len(errs) > 0 {
+		return errs
+	}
+	return nil
+}
+
+// unitTestFile holds the contents of a single unit test file.
+type unitTestFile struct {
+	RuleFiles          []string      `yaml:"rule_files"`
+	EvaluationInterval time.Duration `yaml:"evaluation_interval,omitempty"`
+	GroupEvalOrder     []string      `yaml:"group_eval_order"`
+	Tests              []testGroup   `yaml:"tests"`
+}
+
+func (utf *unitTestFile) maxEvalTime() time.Duration {
+	var maxd time.Duration
+	for _, t := range utf.Tests {
+		d := t.maxEvalTime()
+		if d > maxd {
+			maxd = d
+		}
+	}
+	return maxd
+}
+
+// testGroup is a group of input series and tests associated with it.
+type testGroup struct {
+	Interval        time.Duration    `yaml:"interval"`
+	InputSeries     []series         `yaml:"input_series"`
+	AlertRuleTests  []alertTestCase  `yaml:"alert_rule_test,omitempty"`
+	PromqlExprTests []promqlTestCase `yaml:"promql_expr_test,omitempty"`
+}
+
+// test performs the unit tests.
+func (tg *testGroup) test(mint, maxt time.Time, evalInterval time.Duration, groupOrderMap map[string]int, ruleFiles ...string) []error {
+	// Setup testing suite.
+	suite, err := promql.NewTest(nil, tg.seriesLoadingString())
+	if err != nil {
+		return []error{err}
+	}
+	defer suite.Close()
+
+	err = suite.Run()
+	if err != nil {
+		return []error{err}
+	}
+
+	// Load the rule files.
+	opts := &rules.ManagerOptions{
+		QueryFunc:  rules.EngineQueryFunc(suite.QueryEngine(), suite.Storage()),
+		Appendable: suite.Storage(),
+		Context:    context.Background(),
+		NotifyFunc: func(ctx context.Context, expr string, alerts ...*rules.Alert) {},
+	}
+	m := rules.NewManager(opts)
+	groupsMap, ers := m.LoadGroups(tg.Interval, ruleFiles...)
+	if ers != nil {
+		return ers
+	}
+	groups := orderedGroups(groupsMap, groupOrderMap)
+
+	// Pre-processing some data for testing alerts.
+	// All this preparating so that we can test the alerts as and when we evaluate
```

Comment:
All this preparation is so that we can test alerts as we evaluate the rules. This avoids storing them in memory, as the number of evals might be high.